crates/hir_ty/src/checker/mutability.rs

```diff
@@ -0,0 +1,189 @@
+//! # Mutability & Rebinding
+//!
+//! ## 概要
+//!
+//! Nail は、変数の値が変更されることを防ぐために、変数のミュータビリティと再代入を制御することができます。
+//! ミュータビリティは以下の状態を持ちます。
+//! - 可変
+//! - 不変
+//!
+//! ## 可変性と再代入
+//!
+//! 変数のミュータビリティは、`mut` キーワードを使って指定します。
+//! `mut` キーワードを指定しない場合、変数は不変となります。
+//! 可変な場合、以下の二つを可能にします。
+//! - 変数自体を更新すること
+//! - 再代入を可能にすること
+//!
+//! ```ignore
+//! let mut x = 1;
+//! x = 2; // OK
+//! ```
+//!
+//! ```ignore
+//! let x = 1;
+//! x = 2; // Error
+//! ```
+//!
+//! ## 関数パラメータの可変性(検討中)
+//!
+//! 関数の引数の場合も`mut`をつけることで可変性の許可となります。
+//! 再バインディングは許可されません。再バインディングしたい場合は`let mut`を使用します。
+//!
+//! ```ignore
+//! fn foo(bar: i32) {
+//!   bar = 2; // Error
+//! }
+//! ```
+//!
+//! ```ignore
+//! fn foo(bar: mut i32) {
+//!   bar = 2; // Error
+//! }
+//! ```
+//!
+//! ```ignore
+//! fn foo(bar: mut i32) {
+//!   let mut bar = bar;
+//!   bar = 2; // Ok
+//! }
+//! ```
+
+use std::collections::HashMap;
+
+use crate::TypeCheckError;
+
+pub(crate) struct FunctionMutabilityChecker<'a> {
+    db: &'a dyn hir::HirMasterDatabase,
+    pod: &'a hir::Pod,
+    hir_file: hir::HirFile,
+    function: hir::Function,
+
+    binding_by_expr: HashMap<hir::ExprId, hir::BindingId>,
+    errors: Vec<TypeCheckError>,
+}
+impl<'a> FunctionMutabilityChecker<'a> {
+    pub(crate) fn new(
+        db: &'a dyn hir::HirMasterDatabase,
+        pod: &'a hir::Pod,
+        hir_file: hir::HirFile,
+        function: hir::Function,
+    ) -> Self {
+        Self {
+            db,
+            pod,
+            hir_file,
+            function,
+            binding_by_expr: HashMap::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    pub(crate) fn check(mut self) -> Vec<TypeCheckError> {
+        let hir_file = self.pod.get_hir_file_by_function(self.function).unwrap();
+        let hir::Expr::Block(body) = self.function.body(self.db, *hir_file).unwrap() else {
+            panic!("Function body is not block");
+        };
+
+        for stmt in &body.stmts {
+            self.check_stmt(stmt);
+        }
+        if let Some(tail) = body.tail {
+            self.check_expr(tail);
+        }
+
+        self.errors
+    }
+
+    fn check_stmt(&mut self, stmt: &hir::Stmt) {
+        match stmt {
+            hir::Stmt::Expr {
+                expr,
+                has_semicolon: _,
+            } => self.check_expr(*expr),
+            hir::Stmt::Let { value, binding, .. } => {
+                self.binding_by_expr.insert(*value, *binding);
```

Comment:
**Optimization Issue**: Inserting into 'binding_by_expr' without checking if the key already exists can lead to overwriting existing entries unintentionally, which might cause logical errors in variable tracking. <br> **Fix**: Check if the key exists before inserting to handle potential duplicates or consider updating the logic to accommodate updates to existing entries. <br> **Code Suggestion**: 
 ```
 - self.binding_by_expr.insert(*value, *binding);
 + if !self.binding_by_expr.contains_key(value) {
 +     self.binding_by_expr.insert(*value, *binding);
 + } else {
 +     // Handle duplicate or update logic
 + }
 ```